website/addons/s3/static/tests/s3NodeConfig.test.js

```diff
@@ -0,0 +1,412 @@
+/*global describe, it, expect, example, before, after, beforeEach, afterEach, mocha, sinon*/
+'use strict';
+var $ = require('jquery');
+
+var assert = require('chai').assert;
+
+var utils = require('tests/utils');
+var faker = require('faker');
+
+var s3NodeConfigVM = require('../s3NodeConfig')._S3NodeConfigViewModel;
+
+var API_BASE = '/api/v1/12345/s3';
+var SETTINGS_URL = [API_BASE, 'settings', ''].join('/');
+var URLS = {
+    create_bucket: [API_BASE, 'newbucket', ''].join('/'),
+    import_auth: [API_BASE, 'import-auth', ''].join('/'),
+    create_auth: [API_BASE, 'authorize', ''].join('/'),
+    deauthorize: SETTINGS_URL,
+    bucket_list: [API_BASE, 'buckets', ''].join('/'),
+    set_bucket: SETTINGS_URL,
+    settings: '/settings/addons/'
+};
+var makeSettingsEndpoint = function(result, urls) {
+    return {
+        method: 'GET',
+        url: SETTINGS_URL,
+        response: {
+            result: $.extend({}, {
+                urls: $.extend({}, URLS, urls)
+            }, result)
+        }
+    };
+};
+
+var noop = function() {};
+var APITestCase = function(cfg) {
+    this.description = cfg.description || '';
+    this.endpoint = cfg.endpoint || {
+        response: {
+            result: {}
+        }
+    };
+    this.expected = this.endpoint.response.result;
+    $.extend(this.expected, cfg.data || {});
+};
+APITestCase.prototype.run = function(test) {
+    var tc = this;
+    var server;
+    tc.before = () => {
+        server = utils.createServer(sinon, [tc.endpoint]);
+    };
+    tc.after = () => {
+        server.restore();
+    };
+    test(tc);
+};
+var APITestCases = function(test, cases) {
+    for (var i = 0; i < cases.length; i++) {
+        new APITestCase(cases[i]).run(test);
+    }
+};
+
+describe('s3NodeConfigViewModel', () => {
+    window.contextVars = {
+        node: {
+            urls: {
+                web: '/12345'
+            }
+        }
+    };
+    describe('#fetchFromServer', () => {
+        new APITestCases(
```

Comment:
I'm wondering if this can be simplified by using just nested `describes` and maybe a loop over Objects that contain inputs and expected values. Thoughts?